Remove wait for initialization

arbulu89 opened the pull request:
Remove `wait_for_initialization` parameter. This parameter runs an active sleep to give time to the cluster to finish the initialization in the 1st node. Now that crmsh implements a lock mechanism (https://github.com/ClusterLabs/crmsh/pull/694), this active waiting is not needed, `crmsh` will do it for us.

In order to prevent any further issue, `corosync` configuration file update has been changed, to use the proper tooling instead of restarting the daemon which can create issues. @liangxin1300 Could you confirm that this is a good practice? The exact change in: https://github.com/SUSE/habootstrap-formula/commit/05c2439591215fc84a517044ff7e7d72f4b45345

Finally, I have changed the SSH keys authorization usage to use `ssh-copy-id` instead of implementing this feature.

**Edit:** Pending to test in Azure, which gave most of the problems about this topic in the past
**Edit:** Tested in Azure and worked fine

liangxin1300 commented on cluster/corosync.sls:
```diff
@@ -5,9 +5,9 @@ customize_corosync:
-    - name: corosync
+reload_corosync_configuration:
+  cmd.run:
+    - name: corosync-cfgtool -R
```
@arbulu89 According this https://github.com/corosync/corosync/wiki/Config-file-values
Not all configuration item will be reloaded via `corosync-cfgtool -R`

arbulu89 replied:
Thanks @liangxin1300 ,
Considering this, I will need to do more research. The complete restart of the daemon is needed in some cases, and this affects the formula usage, as a sudden restart of the cluster might affect the other node joining process

arbulu89 replied:
@liangxin1300 Unfortunately some of the variables we update (`secauth` for example) are not covered in `corosync-cfgtool`, so the restart looks needed.
The restart has the big side-effect that stops the cluster completely, so if there is some node joining, this operation will fails.
This makes things more difficult.
In the past, we had a specific `lock` for the formula, which we removed when the lock was implemented in `crmsh`, but this was solving the issue, as we could atomize operations outside of `crmsh`. Now I need to see if there is some other alternative or we need to go back to the lock mechanism, which I don't want.

liangxin1300 replied:
@arbulu89 If there is some node joining, on init node, there should exists `/run/.crmsh_lock_directory`
How about here wait for `/run/.crmsh_lock_directory` removed, then execute restarting corosync.service?

arbulu89 replied:
Yes, good idea. This can be an option. I will need to do some testing because there might be some gap between the joining process creates this file and the corosync service is restarted. To assure everything I would need to create this lock file even in this formula before restarting the corosync service as final resort.
I will keep you updated @liangxin1300

liangxin1300 commented:
@arbulu89 
> Remove wait_for_initialization parameter

Looks good to me, thanks!

nick-wang reviewed (approved):
Good to remove the sleep with lock. 
I think it is worth to bump up to 0.4.2 to align to crmsh internal lock mechanism.

arbulu89 commented:
> Good to remove the sleep with lock.
> I think it is worth to bump up to 0.4.2 to align to crmsh internal lock mechanism.

The versio `0.4.1` is just a coincidence hehe, it doesn't have anything related to crmsh version. I will upgrade the version when the submission is ready though

arbulu89 commented:
Hey @nick-wang and @liangxin1300 ,
I have implemented the next to allow a proper corosync restart: https://github.com/SUSE/habootstrap-formula/pull/76/commits/b5dd1638f40b491ac28032a65fe99b80318c0df1
As commented, the `corosync` service restart is needed. I have implemented again some minimal lock mechanism to assure atomicity on the `init/corosync update` process:

1. The init node creates the lock after the cluster is created. The same lock folder used by crmsh
2. When corosync restart is done, the lock is removed
3. The joining node checks if the lock exists. I know that this 3rd part looks strange, as the crmsh join does the same. But the lock usage in the join of crmsh has a problem: it depends on `crm` commands. When corosync is being restarted, and crmsh tries to check for the lock availability, it runs `crm_mon` (https://github.com/ClusterLabs/crmsh/blob/master/crmsh/lock.py#L158), which fails during the corosync restart. That's why I added this check. 
What do you think guys? I have tested and works. I know that it doesn't look clean, but I don't see more options

arbulu89 commented:
Hey @nick-wang and @liangxin1300 ,
I have implemented the next to allow a proper corosync restart: https://github.com/SUSE/habootstrap-formula/pull/76/commits/b5dd1638f40b491ac28032a65fe99b80318c0df1
As commented, the `corosync` service restart is needed. I have implemented again some minimal lock mechanism to assure atomicity on the `init/corosync update` process:

1. The init node creates the lock after the cluster is created. The same lock folder used by crmsh
2. When corosync restart is done, the lock is removed
3. The joining node checks if the lock exists. I know that this 3rd part looks strange, as the crmsh join does the same. But the lock usage in the join of crmsh has a problem: it depends on `crm` commands. When corosync is being restarted, and crmsh tries to check for the lock availability, it runs `crm_mon` (https://github.com/ClusterLabs/crmsh/blob/master/crmsh/lock.py#L158), which fails during the corosync restart. That's why I added this check. 
What do you think guys? I have tested and works. I know that it doesn't look clean, but I don't see more options

arbulu89 commented:
Closing as it is stalled
